test/e2e/tests/crd-migration.go

```diff
@@ -0,0 +1,215 @@
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+
+		if a, e := crdStorageState.Status.CurrentStorageVersionHash, v1Hash; a != e {
+			util.Failf("unexpected storage version hash %s, expected, %s", a, e)
+		}
+
+		By("Change the storage version of the CRD")
+		output, err = exec.Command("kubectl", "patch", "crd", "tests.migrationtest.k8s.io", `--patch={"spec":{"versions":[{"name":"v1","served":true,"storage":false},{"name":"v2","served":true,"storage":true}]}}`).CombinedOutput()
+		if err != nil {
+			util.Failf("%s", output)
+		}
+
+		By("Wait for the storageVersionHash of the CRD to change in the discovery document")
+		var v2Hash string
+		err = wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
+			v2Hash, err = storageVersionHash(client.Discovery(), "migrationtest.k8s.io/v1", "tests")
+			if err != nil {
+				util.Logf("failed to fetch the storage version of the crd, %v. Retrying.", err)
+				return false, nil
+			}
+			if v2Hash == "" || v2Hash == v1Hash {
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+
+		By("Wait for the storage state of the CRD to change")
+		// Wait for discoveryPeriod + 1 minute to give the triggering controller enough time to detect and react.
+		err = wait.PollImmediate(10*time.Second, discoveryPeriod+1*time.Minute, func() (bool, error) {
+			var err error
+			crdStorageState, err = client.MigrationV1alpha1().StorageStates().Get("tests.migrationtest.k8s.io", metav1.GetOptions{})
+			if err != nil {
+				util.Failf("%v", err)
+			}
+			if crdStorageState.Status.CurrentStorageVersionHash != v2Hash {
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+
+		By("Wait for all storage states to converge")
+		err = wait.PollImmediate(30*time.Second, 10*time.Minute, func() (bool, error) {
+			l, err := client.MigrationV1alpha1().StorageStates().List(metav1.ListOptions{})
+			if err != nil {
+				util.Failf("%v", err)
+			}
+			for _, ss := range l.Items {
+				if len(ss.Status.PersistedStorageVersionHashes) == 1 && ss.Status.PersistedStorageVersionHashes[0] == ss.Status.CurrentStorageVersionHash {
+					continue
+				}
+				util.Logf("resource %v has persisted hashes %v, and current hash %s",
+					ss.Spec.Resource,
+					ss.Status.PersistedStorageVersionHashes,
+					ss.Status.CurrentStorageVersionHash)
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+
+		By("Migrations should have all completed")
+		l, err := client.MigrationV1alpha1().StorageVersionMigrations(namespaceName).List(metav1.ListOptions{})
+		if err != nil {
+			util.Failf("%v", err)
+		}
+		for _, m := range l.Items {
+			if !succeeded(m.Status.Conditions) {
+				util.Failf("unexpected in progress migration for resource %v", m.Spec.Resource)
+			}
+		}
+		// kubectl log into the etcd pod
```

Comment:
what is the comment about?

Reply:
This is a TODO. Added the "TODO:" prefix.